clippy_dev/src/lib.rs

```diff
@@ -56,22 +57,22 @@ impl Lint {
     }
 }
 
-pub fn collect_all() -> Vec<Lint> {
+pub fn gather_all() -> impl Iterator<Item=Lint> {
     let mut lints = vec![];
     for dir_entry in lint_files() {
-        lints.append(&mut collect_from_file(&dir_entry));
+        lints.append(&mut gather_from_file(&dir_entry).collect());
     }
-    lints
+    lints.into_iter()
 }
 
-fn collect_from_file(dir_entry: &fs::DirEntry) -> Vec<Lint> {
+fn gather_from_file(dir_entry: &fs::DirEntry) -> impl Iterator<Item=Lint> {
     let mut file = fs::File::open(dir_entry.path()).unwrap();
     let mut content = String::new();
     file.read_to_string(&mut content).unwrap();
     parse_contents(&content, dir_entry.path().file_stem().unwrap().to_str().unwrap())
 }
 
-fn parse_contents(content: &str, filename: &str) -> Vec<Lint> {
+fn parse_contents(content: &str, filename: &str) -> impl Iterator<Item=Lint> {
     let mut lints: Vec<Lint> = DEC_CLIPPY_LINT_RE
```

Comment:
chain the two iterators and remove the collect&append&into_iter

Reply:
If I don't collect the map result, I run into a lot of lifetime issues, because `contents` and `filename` are used in the iterator and would outlive the function. I will push the changes in a bit with a comment.